File: python/taflow/metrics/gross_loss.py
"""Signed gross-loss metric for raw P&L observations."""

from __future__ import annotations

from typing import Any

from .._native.metrics import GrossLoss as _Native
from ._input import as_metric_series


class GrossLoss:
    """Compute the signed sum of strictly negative raw P&L values.

    Negative observations are added without changing their sign; positive and
    exact-zero observations contribute zero. Warm-up requires one usable P&L
    observation. Empty input returns ``None``, while a non-empty history with
    no loss validly returns ``0.0``. The independent oracle is NumPy's strict
    negative filtered sum. QuantStats 0.0.81 ``profit_factor`` provides an
    executable cross-check through the absolute value of the same denominator
    when called with ``prepare_returns=False``.

    ``from_pnl`` consumes non-cumulative raw period P&L and ``from_trades``
    consumes realized P&L from closed trades. Both preserve monetary values
    exactly and accept no initial-equity argument. Return, equity, and
    log-return factories are deliberately unavailable because Gross Loss is an
    absolute P&L statistic. Rust owns O(1)-memory arithmetic. NaNs are omitted
    by default or rejected with ``nan_policy="raise"``; infinities are always
    rejected. Mutating lifecycle methods are fluent and bulk work releases the
    GIL.
    """

    def __init__(self) -> None:
        """Reject ambiguous construction; use a semantic ``from_*`` factory."""
        raise TypeError("use GrossLoss.from_pnl/from_trades")

    @classmethod
    def _create(
        cls,
        values: Any,
        input_mode: str,
        *,
        nan_policy: str,
        column: str | None,
    ) -> "GrossLoss":
        state = cls.__new__(cls)
        state._state = _Native(input_mode, nan_policy)
        return state.extend(values, column=column)

    @classmethod
    def from_pnl(
        cls,
        pnl: Any,
        *,
        nan_policy: str = "omit",
        column: str | None = None,
    ) -> "GrossLoss":
        """Construct from raw non-cumulative period P&L without conversion."""
        return cls._create(pnl, "pnl", nan_policy=nan_policy, column=column)

    @classmethod
    def from_trades(
        cls,
        trades: Any,
        *,
        nan_policy: str = "omit",
        column: str | None = None,
    ) -> "GrossLoss":
        """Construct from realized P&L for chronological closed trades."""
        return cls._create(trades, "trades", nan_policy=nan_policy, column=column)

    def append(self, value: float) -> "GrossLoss":
        """Append one raw P&L observation and return this metric."""
        self._state.append(float(value))
        return self

    def extend(
        self, values: Any, *, column: str | None = None
    ) -> "GrossLoss":
        """Append chronological raw P&L observations and return this metric."""
        self._state.extend(as_metric_series(values, column=column))
        return self

    @property
    def value(self) -> float | None:
        """Return signed gross loss, or ``None`` while empty."""
        return self._state.value

    def compute(self) -> float | None:
        """Return the current scalar without replaying processed P&L."""
        return self._state.compute()

    def reset(self) -> "GrossLoss":
        """Clear observations, preserve the P&L domain, and return this metric."""
        self._state.reset()
        return self

    def __len__(self) -> int:
        """Return the number of usable P&L observations processed by Rust."""
        return len(self._state)


__all__ = ["GrossLoss"]

File: crates/taflow-metrics/src/metrics/gross_loss.rs
use crate::{
    primitives::GainLossState, MetricError, MetricInputKind, MetricInputState, MetricResult,
    NanPolicy,
};

/// Signed sum of strictly negative raw P&L observations.
#[derive(Debug, Clone)]
pub struct GrossLoss {
    input: MetricInputState,
    observations: GainLossState,
}

impl GrossLoss {
    /// Construct an empty state for raw period P&L or realized closed-trade P&L.
    pub fn new(input_kind: MetricInputKind, nan_policy: NanPolicy) -> MetricResult<Self> {
        if !matches!(
            input_kind,
            MetricInputKind::RawPnl | MetricInputKind::Trades
        ) {
            return Err(MetricError::InvalidParameter {
                name: "input_kind",
                value: format!("{input_kind:?}"),
                reason: "gross loss requires raw period P&L or realized closed-trade P&L",
            });
        }
        Ok(Self {
            input: MetricInputState::new(input_kind, nan_policy)?,
            observations: GainLossState::new(),
        })
    }

    /// Append one chronological P&L observation and return signed gross loss.
    pub fn append(&mut self, value: f64) -> MetricResult<Option<f64>> {
        if let Some(observation) = self.input.append(value)? {
            self.observations.append(observation);
        }
        Ok(self.value())
    }

    /// Append a chronological slice through the same persistent state.
    pub fn extend(&mut self, values: &[f64]) -> MetricResult<Option<f64>> {
        self.input.extend(values, |observation| {
            self.observations.append(observation);
            Ok(())
        })?;
        Ok(self.value())
    }

    /// Return signed gross loss, or `None` when no usable P&L was processed.
    pub fn value(&self) -> Option<f64> {
        (!self.input.is_empty()).then(|| self.observations.gross_loss())
    }

    /// Return the current result without replaying prior observations.
    pub fn compute(&self) -> Option<f64> {
        self.value()
    }

    /// Restore fresh-state behavior while preserving the P&L input domain.
    pub fn reset(&mut self) {
        self.input.reset();
        self.observations.reset();
    }

    /// Return the number of usable P&L observations processed.
    pub fn len(&self) -> usize {
        self.input.len()
    }

    /// Return whether no usable observations have been processed.
    pub fn is_empty(&self) -> bool {
        self.input.is_empty()
    }
}
